tfx/dsl/components/base/base_beam_component.py

```diff
@@ -46,6 +46,22 @@ def with_beam_pipeline_args(
          self.executor_spec).add_beam_pipeline_args(beam_pipeline_args)
     return self
 
+  def with_beam_pipeline_args_from_env(
+      self, beam_pipeline_args_from_env: Dict[str, str]) -> 'BaseBeamComponent':
```

Comment:
Not sure where this code is duplicated, can you specify where?

The merging behaviour is the same as for ordinary beam args, i.e. component specific args override pipeline level args.

And in addition, any arg specified as an ordinary beam arg will take precedence over any kind of beam arg from env.

Reply:
I think that the merging behavior seems clear now thanks to the explanation in the proto. Thanks!

But I am a little bit worried about the duplication of "beam_pipeline_args" and "beam_pipeline_args_from_env" in general. Because these two things appears together almost always. From the first comment by @zhitaoli (https://github.com/tensorflow/tfx/pull/3455#issuecomment-825155321 ) we might need to explore ideas using some kind of placeholders. If the existing placeholder in TFX doesn't fit well here, we might be able to invent a new simple placeholder mechanism like shell env var expansion.

We might be able to proceed with the current implementation, but I think that we might need to get more inputs on this from other folks. (CC @kennethyang404 )